flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageProducerClient.java

```diff
@@ -0,0 +1,97 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage;
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferCompressor;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageSubpartitionId;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierProducerAgent;
+import org.apache.flink.util.ExceptionUtils;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/** Client of the Tiered Storage used by the producer. */
+public class TieredStorageProducerClient {
+    private final boolean isBroadcastOnly;
+
+    private final int numSubpartitions;
+
+    private final BufferAccumulator bufferAccumulator;
+
+    private final BufferCompressor bufferCompressor;
+
+    private final List<TierProducerAgent> tierProducerAgents;
+
+    public TieredStorageProducerClient(
+            int numSubpartitions,
+            boolean isBroadcastOnly,
+            BufferAccumulator bufferAccumulator,
+            @Nullable BufferCompressor bufferCompressor,
+            List<TierProducerAgent> tierProducerAgents) {
+        this.isBroadcastOnly = isBroadcastOnly;
+        this.numSubpartitions = numSubpartitions;
+        this.bufferAccumulator = bufferAccumulator;
+        this.bufferCompressor = bufferCompressor;
+        this.tierProducerAgents = tierProducerAgents;
+
+        bufferAccumulator.setup(numSubpartitions, this::writeFinishedBuffers);
+    }
+
+    public void write(
+            ByteBuffer record,
+            TieredStorageSubpartitionId subpartitionId,
+            Buffer.DataType dataType,
+            boolean isBroadcast)
+            throws IOException {
+
+        if (isBroadcast && !isBroadcastOnly) {
+            for (int i = 0; i < numSubpartitions; ++i) {
+                bufferAccumulator.receive(record.duplicate(), subpartitionId, dataType);
+            }
+        } else {
+            bufferAccumulator.receive(record, subpartitionId, dataType);
+        }
+    }
+
+    public void close() {
+        bufferAccumulator.close();
+        tierProducerAgents.forEach(TierProducerAgent::close);
+    }
+
+    public void writeFinishedBuffers(
```

Comment:
This should be private. And I'd suggest the name `writeAccumulatedBuffers`. Otherwise, we need to explain what is the concept `finished` for a buffer.

Reply:
Renemed it.